[BPF] Visit nested map array during BTF generation

mtardy opened the pull request:
Fixes missing inner map struct type definitions [^1]. We should visit the type of nested array of maps like we do for global maps. This patch adds a boolean to convey the information to visitTypeEntry and visitDerivedType that the pointee is a map definition and should be treated as such.

It ressembles and works with commit 0d21c956a5c1 ("[BPF] Handle nested wrapper structs in BPF map definition traversal (#144097)") which focused on directly nested wrapper structs.

Before that patch, this ARRAY_OF_MAPS definition would lead to the BTF information include the 'missing_type' as "FWD 'missing_type' fwd_kind=struct":

	struct missing_type { uint64_t foo; };
	struct {
		__uint(type, BPF_MAP_TYPE_ARRAY_OF_MAPS);
		[...]
		__array(
			values, struct {
				[...]
				__type(value, struct missing_type);
			});
	} map SEC(".maps");

Which lead to errors while trying to load the map:

	libbpf: map 'outer_map.inner': can't determine value size for type [N]: -22.

To solve this issue, users had to use the struct in a dummy variable or in a dummy function for the BTF to be generated correctly [^2].

[^1]: https://lore.kernel.org/netdev/aH_cGvgC20iD8qs9@gmail.com/T/#u
[^2]: https://github.com/cilium/ebpf/discussions/1658#discussioncomment-12491339

mtardy committed:
[BPF] Visit nested map array during BTF generation

Fixes missing inner map struct type definitions [^1]. We should visit
the type of nested array of maps like we do for global maps. We visit
the derived type of the array directly in visitMapDefType.

Because of the small changes in visitMapDefType, ordering of the BTF is
changed, for example, the PTR for the nested maps might appear later.
This explain why this patch had to update the BTF checks of the tests
'map-def-nested-array.ll', 'map-def-2.ll' and 'map-def-3.ll'. We took
the opportunity to migrate the last two to use 'print_btf.py'.

It ressembles commit 0d21c956a5c1 ("[BPF] Handle nested wrapper structs
in BPF map definition traversal (#144097)") which focused on directly
nested wrapper structs.

Before that patch, this ARRAY_OF_MAPS definition would lead to the BTF
information include the 'missing_type' as "FWD 'missing_type'
fwd_kind=struct":

	struct missing_type { uint64_t foo; };
	struct {
		__uint(type, BPF_MAP_TYPE_ARRAY_OF_MAPS);
		[...]
		__array(
			values, struct {
				[...]
				__type(value, struct missing_type);
			});
	} map SEC(".maps");

Which lead to errors while trying to load the map:

	libbpf: map 'outer_map.inner': can't determine value size for type [N]: -22.

To solve this issue, users had to use the struct in a dummy variable or
in a dummy function for the BTF to be generated correctly [^2].

[^1]: https://lore.kernel.org/netdev/aH_cGvgC20iD8qs9@gmail.com/T/#u
[^2]: https://github.com/cilium/ebpf/discussions/1658#discussioncomment-12491339

Co-authored-by: Eduard Zingerman <eddyz87@gmail.com>
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

eddyz87 commented:
This whole business with boolean flags for `visitDerivedType` gets out of hand.
I need to think a bit.

yonghong-song commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -707,7 +707,7 @@ void BTFDebug::visitArrayType(const DICompositeType *CTy, uint32_t &TypeId) {
   uint32_t ElemTypeId;
   const DIType *ElemType = CTy->getBaseType();
-  visitTypeEntry(ElemType, ElemTypeId, false, false);
+  visitTypeEntry(ElemType, ElemTypeId, false, false, false);
```
Indeed, now we have three bool parameters. Probably we should have a struct to replace these bool parameters now.

mtardy commented:
> Probably we should have a struct to replace these bool parameters now

If you want I can do a follow up patch in this PR with this. Or in another.

eddyz87 commented:
Please wait a bit.

eddyz87 committed:
[BPF] Reorganize BTFDebug::visitMapDefType for better readability

Kernel BPF selftests passing. No change in BTF generated for kernel
selftests, modulo some reorderings.

Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

eddyz87 commented:
First, I think that there is no need to modify `visitTypeEntry` at all, the changes can be confined to `visitMapDefType`, as in [this commit](https://github.com/llvm/llvm-project/commit/87326998aecd464723822a7429b6aaa9ff708be5).

Second, I think that `visitMapDefType` should be refactored to make intent more clear, as in [this commit](https://github.com/llvm/llvm-project/commit/86838de43e58ba733e6e691f3859da293c555949).

In both cases llvm unit tests require adjustments, kernel BPF selftests are passing, no significant changes in BPF generated for kernel BPF selftests.

yonghong-song commented:
That is true. If in visitMapDefType() we can identify inner maps, we can just recurse visitMapDefType() itself. This should make code simpler.

mtardy commented:
> First, I think that there is no need to modify `visitTypeEntry` at all, the changes can be confined to `visitMapDefType`, as in [this commit](https://github.com/llvm/llvm-project/commit/87326998aecd464723822a7429b6aaa9ff708be5).

Ah nice, I thought we needed to `visitDerivedType` anyway to generated the BTF type for the pointers to the map type (so I wrote such patch, skipping just `visitTypeEntry`), but it seems the `visitTypeEntry` at the end of `visitMapDefType` does the job (in a different order, thus the unit test BTF output modification). Thank you!

> Second, I think that `visitMapDefType` should be refactored to make intent more clear, as in [this commit](https://github.com/llvm/llvm-project/commit/86838de43e58ba733e6e691f3859da293c555949).

Agree, it's better this way, only logical difference is you don't return on `CTy->isForwardDecl()`. I imagine you noticed and it's okay but just saying.

https://github.com/llvm/llvm-project/blob/2f2df751d453566ab70fd02b9a019cd66af76bc6/llvm/lib/Target/BPF/BTFDebug.cpp#L976-L977

eddyz87 commented:
> Agree, it's better this way, only logical difference is you don't return on `CTy->isForwardDecl()`. I imagine you noticed and it's okay but just saying.
> 
> https://github.com/llvm/llvm-project/blob/2f2df751d453566ab70fd02b9a019cd66af76bc6/llvm/lib/Target/BPF/BTFDebug.cpp#L976-L977

I thought about it and concluded that it shouldn't matter with the following reasoning: for top-level or nested structures in map key/value definition the reference in debug info shouldn't be a forward declaration. Even if it happens to be somehow, old code didn't handle this situation gracefully (didn't search for corresponding definition).

eddyz87 commented:
@mtardy ,

Could you please fix the following tests:

```
Failed Tests (3):
  LLVM :: CodeGen/BPF/BTF/map-def-2.ll
  LLVM :: CodeGen/BPF/BTF/map-def-3.ll
  LLVM :: CodeGen/BPF/BTF/map-def-nested-array.ll
```

Some reordering is because of my changes in patch #1?
Probably would be nice to switch `map-def-{2,3}.ll` to `./llvm/test/CodeGen/BPF/BTF/print_btf.py`.

mtardy commented:
> Could you please fix the following tests:
> 
> ```
> Failed Tests (3):
>   LLVM :: CodeGen/BPF/BTF/map-def-2.ll
>   LLVM :: CodeGen/BPF/BTF/map-def-3.ll
>   LLVM :: CodeGen/BPF/BTF/map-def-nested-array.ll
> ```
> 
> Some reordering is because of my changes in patch #1? Probably would be nice to switch `map-def-{2,3}.ll` to `./llvm/test/CodeGen/BPF/BTF/print_btf.py`.

Ah yep I missed this indeed! This should be fixed now and I migrated the tests to use `print_btf.py`.
